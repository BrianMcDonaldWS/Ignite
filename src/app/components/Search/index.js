import React, { Component } from 'react';
import PropTypes from 'prop-types';
import makeClass from 'classnames';
import { throttle } from 'throttle-debounce';
import SearchApi from 'js-worker-search';
import getLineNumber from 'get-line-from-pos';
import replaceAt from '../../../utils/replace-at';

import styles from './search.css';

const indexOfAll = (source, term) => {
  const indexes = [];

  let index = source.toLowerCase().indexOf(term);

  while (index !== -1) {
    indexes.push(index);
    index = source.indexOf(term, index + 1);
  }

  return indexes;
};

const lineWithCodeBlock = (line, term) => {
  let tickIndex = line.indexOf('`');
  let termIndex = line.indexOf(term);

  while (termIndex !== -1) {
    if (termIndex < tickIndex || tickIndex === -1) {
      line = replaceAt(line, term, `**${term}**`, termIndex);
      termIndex = line.indexOf(term, termIndex + 3);

      if (tickIndex !== -1) {
        tickIndex += 4; // Account for **
      }
    } else if (line.indexOf('`', tickIndex + 1) < termIndex) {
      const paired = line.indexOf('`', tickIndex + 1);

      if (paired === -1) {
        tickIndex = line.indexOf('`', tickIndex + 1);
      } else {
        tickIndex = line.indexOf('`', paired + 1);
      }
    } else {
      termIndex = line.indexOf(term, termIndex + 1);
    }
  }

  return line;
};

function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

const getLines = (source, indexes, term) => {
  return new Set(
    indexes.map(index => {
      const lineNumber = getLineNumber(source, index);
      let line = source.split('\n')[lineNumber - 1];

      if (line.indexOf('`') > -1) {
        line = lineWithCodeBlock(line, term);
      } else if (line.indexOf('![') === -1) {
        line = line.replace(new RegExp(`${term}`), `**${term}**`);
        line = line.replace(
          new RegExp(`${capitalizeFirstLetter(term)}`, 'g'),
          `**${capitalizeFirstLetter(term)}**`
        );
      }

      return line;
    })
  );
};

class Search extends Component {
  static propTypes = {
    setSearchResults: PropTypes.func
  };

  static defaultProps = {
    setSearchResults: () => {}
  };

  constructor(props) {
    super(props);

    this.index = new SearchApi();
    this.constructIndex();
  }

  constructIndex = () => {
    this.props.searchIndex.forEach(file => {
      this.index.indexDocument(file.id, file.body);
    });
  };

  componentDidUpdate = () => {
    this.constructIndex();
  };

  search = throttle(500, async term => {
    if (term === '') {
      return this.props.setSearchResults([]);
    }

    let results = await this.index.search(term);
    results = results.map(result => {
      const page = this.props.searchIndex.find(file => file.id === result);
      const indexes = indexOfAll(page.body, term);

      return [page.id, getLines(page.body, indexes, term)];
    });

    this.props.setSearchResults(results);
  });

  keyDown = event => {
    if (event.keyCode === 13) {
      this.search(event.target.value);
    }

    if (event.keyCode === 8) {
      this.search(event.target.value);
    }
  };

  render() {
    return process.env.static ? null : (
      <div className={makeClass('navbar-item', styles.root)}>
        <div className="control">
          <input
            className="input"
            type="text"
            placeholder="Search"
            onChange={e => this.search(e.target.value)}
            onKeyDown={this.keyDown}
          />
        </div>
      </div>
    );
  }
}

export default Search;
