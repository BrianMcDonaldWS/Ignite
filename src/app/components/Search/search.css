.root {
  input,
  ::-webkit-input-placeholder {
    text-align: right;
  }
}
