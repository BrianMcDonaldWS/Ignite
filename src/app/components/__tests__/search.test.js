import React from 'react';

import Search, { indexOfAll, lineWithCodeBlock, getLines } from '../Search';
import renderToJson from './utils/render-to-json';

describe('Search', () => {
  test('should render a search', () => {
    expect(
      renderToJson(
        <Search
          searchIndex={[
            { id: 'page.md', body: 'all the markdown text for page.md' }
          ]}
        />
      )
    ).toMatchSnapshot();
  });

  test('should not render a search in static mode', () => {
    expect(
      renderToJson(
        <Search
          static
          searchIndex={[
            { id: 'page.md', body: 'all the markdown text for page.md' }
          ]}
        />
      )
    ).toBeNull();
  });
});

test('indexOfAll', () => {
  expect(
    indexOfAll(
      'asdf adsf asdf foo asdfkasjfdlasd foo asdjfkasjdf fo asdjfajsdf foo asjdfajsdf',
      'foo'
    )
  ).toEqual([15, 34, 64]);
});

test('lineWithCodeBlock', () => {
  expect(
    lineWithCodeBlock('foo ``` foo bar baz foo ``` `foo` `foo bar foo ', 'foo')
  ).toBe('**foo** ``` foo bar baz foo ``` `foo` `**foo** bar **foo** ');
});

test('getLines', () => {
  const term = 'foo';
  const source = `
    Some line without any term.

    foo \`foo\`

    more 
    lines

    foobar baz.
  `;
  const indexes = indexOfAll(source, term);
  expect(getLines(source, indexes, term)).toEqual(
    new Set(['    **foo** `foo`', '    **foo**bar baz.'])
  );
});
