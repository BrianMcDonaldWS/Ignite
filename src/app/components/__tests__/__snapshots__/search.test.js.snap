// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render a search 1`] = `
<div
  className="navbar-item root"
>
  <div
    className="control"
  >
    <input
      className="input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Search"
      type="text"
    />
  </div>
</div>
`;
